service/kvm_config_framework.go

```diff
@@ -0,0 +1,195 @@
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+func newKVMConfigFramework(config Config) (*framework.Framework, error) {
+	if config.Flag == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Flag must not be empty")
+	}
+	if config.Viper == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Viper must not be empty")
+	}
+
+	var err error
+
+	var restConfig *rest.Config
+	{
+		c := k8srestconfig.DefaultConfig()
+
+		c.Logger = config.Logger
+
+		c.Address = config.Viper.GetString(config.Flag.Service.Kubernetes.Address)
+		c.InCluster = config.Viper.GetBool(config.Flag.Service.Kubernetes.InCluster)
+		c.TLS.CAFile = config.Viper.GetString(config.Flag.Service.Kubernetes.TLS.CAFile)
+		c.TLS.CrtFile = config.Viper.GetString(config.Flag.Service.Kubernetes.TLS.CrtFile)
+		c.TLS.KeyFile = config.Viper.GetString(config.Flag.Service.Kubernetes.TLS.KeyFile)
+
+		restConfig, err = k8srestconfig.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	clientSet, err := versioned.NewForConfig(restConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	k8sExtClient, err := apiextensionsclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var crdClient *k8scrdclient.CRDClient
+	{
+		c := k8scrdclient.DefaultConfig()
+
+		c.K8sExtClient = k8sExtClient
+		c.Logger = config.Logger
+
+		crdClient, err = k8scrdclient.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var certsSearcher certs.Interface
+	{
+		c := certs.DefaultConfig()
+
+		c.K8sClient = k8sClient
+		c.Logger = config.Logger
+
+		certsSearcher, err = certs.NewSearcher(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var randomkeysSearcher randomkeys.Interface
+	{
+		keyConfig := randomkeys.DefaultConfig()
+		keyConfig.K8sClient = k8sClient
+		keyConfig.Logger = config.Logger
+		randomkeysSearcher, err = randomkeys.NewSearcher(keyConfig)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var resourcesV2 []framework.Resource
+	{
+		c := defaultKVMConfigResourcesV2Config()
+
+		c.CertsSearcher = certsSearcher
+		c.K8sClient = k8sClient
+		c.Logger = config.Logger
+		c.RandomkeysSearcher = randomkeysSearcher
+
+		c.Name = config.Name
+
+		resourcesV2, err = newKVMConfigResourcesV2(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var resourcesV3 []framework.Resource
+	{
+		c := defaultKVMConfigResourcesV3Config()
+
+		c.CertsSearcher = certsSearcher
+		c.K8sClient = k8sClient
+		c.Logger = config.Logger
+		c.RandomkeysSearcher = randomkeysSearcher
+
+		c.Name = config.Name
+
+		resourcesV3, err = newKVMConfigResourcesV3(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
```

Comment:
resroucesV2 and resourcesV3 initialization is just a function call now.

Reply:
Nice!